### PR TITLE
balancer/rls: Add picker and cache unit tests for RLS Metrics

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds picker and cache unit tests to test RLS Metrics.

RELEASE NOTES: N/A

### CODE REVIEW

File: internal/testutils/stats/test_metrics_recorder.go [hunk 1]
```diff
@@ -75,6 +75,15 @@ func (r *TestMetricsRecorder) AssertDataForMetric(metricName string, wantVal flo
 	}
 }
 
+// AssertNoDataForMetric asserts no data is present for metric.
+func (r *TestMetricsRecorder) AssertNoDataForMetric(metricName string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if _, ok := r.data[estats.Metric(metricName)]; ok {
+		r.t.Fatalf("Data is present for metric %v", metricName)
```
Comment (dfawley) [inline internal/testutils/stats/test_metrics_recorder.go:83]: This is following an anti-pattern: https://google.github.io/styleguide/go/decisions#assert

Now that we have more time, let's address this.

File: balancer/rls/cache_test.go [hunk 2]
```diff
@@ -242,3 +242,50 @@ func (s) TestDataCache_ResetBackoffState(t *testing.T) {
 		t.Fatalf("unexpected diff in backoffState for cache entry after dataCache.resetBackoffState(): %s", diff)
 	}
 }
+
+var cacheEntriesMetricsTests = []*cacheEntry{
```
Comment (dfawley) [inline balancer/rls/cache_test.go:246]: Please move inside test function so it's not global.

File: balancer/rls/cache_test.go [hunk 3]
```diff
@@ -242,3 +242,50 @@ func (s) TestDataCache_ResetBackoffState(t *testing.T) {
...
+	// 5 total entries, size 3 each, so should record 5 entries and 15
+	// size.
+	tmr.AssertDataForMetric("grpc.lb.rls.cache_entries", 5)
+	tmr.AssertDataForMetric("grpc.lb.rls.cache_size", 15)
+
+	// Resize down the cache to 3 entries. This should scale down the cache to 3
+	// entries with 3 bytes each, so should record 3 entries and 9 size.
+	dc.resize(9)
+	tmr.AssertDataForMetric("grpc.lb.rls.cache_entries", 3)
+	tmr.AssertDataForMetric("grpc.lb.rls.cache_size", 9)
```
Comment (dfawley) [inline balancer/rls/cache_test.go:272]: This relies on having the same size for every entry? Or does order matter?

File: balancer/rls/cache_test.go [hunk 4]
```diff
@@ -242,3 +242,50 @@ func (s) TestDataCache_ResetBackoffState(t *testing.T) {
 		t.Fatalf("unexpected diff in backoffState for cache entry after dataCache.resetBackoffState(): %s", diff)
 	}
 }
+
+var cacheEntriesMetricsTests = []*cacheEntry{
+	{size: 3},
+	{size: 3},
+	{size: 3},
+	{size: 3},
```
Comment (dfawley) [inline balancer/rls/cache_test.go:250]: It seems like these should vary to have better coverage.

File: balancer/rls/cache_test.go (hunk 2 above)
Comment (zasweq) [reply balancer/rls/cache_test.go:246]: Ah good point. Done.

File: balancer/rls/cache_test.go (hunk 3 above)
Comment (zasweq) [reply balancer/rls/cache_test.go:272]: Yeah it relies. I think operations are deterministic if I tweak the numbers, but I don't know if we'll have that guarantee in the future. Let me play around with this some.

File: balancer/rls/cache_test.go (hunk 4 above)
Comment (zasweq) [reply balancer/rls/cache_test.go]: Switched to 12345, and scaled the assertions to follow this. Luckily all the cache operations are deterministic.

File: internal/testutils/stats/test_metrics_recorder.go (hunk 1 above)
Comment (zasweq) [reply internal/testutils/stats/test_metrics_recorder.go]: Didn't pass t to this component anymore, did all the assertions inline.

File: balancer/weightedroundrobin/balancer_test.go [hunk 5]
```diff
@@ -234,12 +235,28 @@ func (s) TestWRRMetricsBasic(t *testing.T) {
...
+	}
+	if got := mr.Data[estats.Metric("grpc.lb.wrr.endpoint_weight_not_yet_usable")]; got != 1 {
+		t.Fatalf("Unexpected data for metric %v, got: %v, want: %v", "grpc.lb.wrr.endpoint_weight_not_yet_usable", got, 1)
+	}
+	// Unusable, so no endpoint weight. Due to only one SubConn, this will never
+	// update the weight. Thus, this will stay 0.
+	if got := mr.Data[estats.Metric("grpc.lb.wrr.endpoint_weight_stale")]; got != 0 {
+		t.Fatalf("Unexpected data for metric %v, got: %v, want: %v", "grpc.lb.wrr.endpoint_weight_stale", got, 0)
+	}
+	/*mr.AssertDataForMetric("grpc.lb.wrr.rr_fallback", 1)           // Falls back because only one SubConn.
```
Comment (dfawley) [inline balancer/weightedroundrobin/balancer_test.go:254]: Please fix this up.

File: balancer/rls/cache_test.go [hunk 6]
```diff
@@ -242,3 +243,61 @@ func (s) TestDataCache_ResetBackoffState(t *testing.T) {
...
+	for i, k := range cacheKeys {
+		dc.addEntry(k, cacheEntriesMetricsTests[i])
+	}
+
+	const cacheEntriesKey = "grpc.lb.rls.cache_entries"
+	const cacheSizeKey = "grpc.lb.rls.cache_size"
+	// 5 total entries which add up to 15 size, so should record that.
+	if got := tmr.Data[estats.Metric(cacheEntriesKey)]; got != 5 {
+		t.Fatalf("Unexpected data for metric %v, got: %v, want: %v", cacheEntriesKey, got, 5)
+	}
```
Comment (dfawley) [inline balancer/rls/cache_test.go:268]: Consider the pattern: `if got, want := tmr.Data[], 5; got != want { t.Fatalf(..., got, want) }`

File: balancer/weightedroundrobin/balancer_test.go [hunk 7]
```diff
@@ -87,7 +88,7 @@ var (
 		OOBReportingPeriod:      stringp("0.005s"),
 		BlackoutPeriod:          stringp("0s"),
 		WeightExpirationPeriod:  stringp("60s"),
-		WeightUpdatePeriod:      stringp(".050s"),
+		WeightUpdatePeriod:      stringp("30s"),
```
Comment (dfawley) [inline balancer/weightedroundrobin/balancer_test.go:91]: Why is this changed?

File: internal/testutils/stats/test_metrics_recorder.go [hunk 8]
```diff
@@ -44,57 +44,30 @@ type TestMetricsRecorder struct {
 	floatHistoCh *testutils.Channel
 	intGaugeCh   *testutils.Channel
 
-	// The most recent update for each metric name.
-	mu   sync.Mutex
-	data map[estats.Metric]float64
+	// Mu protects Data.
+	Mu sync.Mutex
+	// Data is the most recent update for each metric name.
+	Data map[estats.Metric]float64
```
Comment (dfawley) [inline internal/testutils/stats/test_metrics_recorder.go:50]: A getter may be better than exporting?

```go
func (...) Metric(name string) float64
```

File: balancer/weightedroundrobin/balancer_test.go (hunk 7 above)
Comment (zasweq) [reply balancer/weightedroundrobin/balancer_test.go:91]: Not strictly needed; but I figured I'd make it hard for the scheduler to emit during the test which would race with test (but I protect with mu so no issue there).

File: balancer/rls/cache_test.go (hunk 6 above)
Comment (zasweq) [reply balancer/rls/cache_test.go:268]: Yeah I was doing everything not aliased and the gots were too long and figured I could do it with wants to, but decided just to do got. It looks nice though so I'l switch this.

File: balancer/weightedroundrobin/balancer_test.go (hunk 5 above)
Comment (zasweq) [reply balancer/weightedroundrobin/balancer_test.go:254]: Whoops deleted.

File: balancer/rls/cache_test.go (hunk 6 above)
Comment (zasweq) [reply balancer/rls/cache_test.go:268]: Discussed offline; not possible due to returning two values from helper.

File: internal/testutils/stats/test_metrics_recorder.go [hunk 9]
```diff
@@ -36,18 +35,16 @@ import (
...
-
 	intCountCh   *testutils.Channel
 	floatCountCh *testutils.Channel
 	intHistoCh   *testutils.Channel
 	floatHistoCh *testutils.Channel
 	intGaugeCh   *testutils.Channel
 
-	// Mu protects Data.
+	// Mu protects data.
 	Mu sync.Mutex
```
Comment (dfawley) [inline internal/testutils/stats/test_metrics_recorder.go:45]: The main goal of making a getter was so that `mu` didn't need to be exported and to make access simpler, since mutexes are generally internal state and complicate things.

Seems like you need something more complex if you need an atomic read of multiple values (though this wasn't supported previously?).  `func Metrics(name []string) (*float64)`???

Comment (zasweq) [reply internal/testutils/stats/test_metrics_recorder.go:45]: Hmmmmm I just deleted the grab of it in balancer_test. I was worried due to the low time for weight updates (when time.Ticker goes off and reupdates scheduler) but I made it 30 seconds so should never race (it'll fail much earlier than that). So mutex still grabbed sometimes in this component in exported helpers.

**Review by dfawley — APPROVED** [review]
